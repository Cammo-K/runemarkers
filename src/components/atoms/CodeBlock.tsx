import { Tile } from '@/types';
import { useCallback, useEffect, useState } from 'react';
import styles from './CodeBlock.module.css';

const TRUNCATE_LENGTH = 500;
const TRUNCATE_LINES = 10;

const getPosition = (string: string, subString: string, index: number) => {
  return string.split(subString, index).join(subString).length;
};

export default function CodeBlock({
  tiles,
  truncateLength,
}: {
  tiles: Tile[];
  truncateLength?: number;
}) {
  const [pretty, setPretty] = useState<boolean>(false);

  const getTruncateNum = useCallback(() => {
    if (truncateLength) return truncateLength;
    if (pretty) return TRUNCATE_LINES;
    return TRUNCATE_LENGTH;
  }, [truncateLength, pretty]);

  const [truncateNum, setTruncateNum] = useState<number>(getTruncateNum());

  const getText = useCallback(
    () =>
      pretty
        ? JSON.stringify(tiles, null, `\t`).toString()
        : JSON.stringify(tiles).toString(),
    [tiles, pretty],
  );

  const [text, setText] = useState<string>(getText());

  const getLength = useCallback(
    () => (pretty ? (text.match(/\n/g) || []).length : text.length),
    [text, pretty],
  );

  const [length, setLength] = useState<number>(getLength());

  const checkTruncated = useCallback(() => {
    return length > truncateNum;
  }, [length, truncateNum]);

  const [truncated, setTruncated] = useState<boolean>(checkTruncated());

  const getTruncatedPos = useCallback(
    () => (pretty ? getPosition(text, `\n`, truncateNum) : truncateNum),
    [text, pretty, truncateNum],
  );

  const [truncatedPos, setTruncatedPos] = useState<number>(getTruncatedPos());

  useEffect(() => {
    setTruncated(checkTruncated());
    setTruncateNum(getTruncateNum());
    setText(getText());
    setTruncatedPos(getTruncatedPos());
    setLength(getLength());
  }, [getText, getTruncateNum, getLength, checkTruncated, getTruncatedPos]);

  const [showingAll, setShowingAll] = useState(false);

  return (
    <div className={styles.container}>
      <h2 className={styles.header}>Tile Data</h2>
      <div className={styles.options}>
        {truncated ? (
          <code
            className={[
              styles.toggleShowAll,
              showingAll ? styles.active : ``,
            ].join(` `)}
            onClick={() => {
              setShowingAll(!showingAll);
            }}
          >
            {showingAll
              ? `show less`
              : `show all (${length - truncateNum} more ${
                  pretty ? `lines` : `characters`
                })`}
          </code>
        ) : null}
        <code
          className={[styles.togglePretty, pretty ? styles.active : ``].join(
            ` `,
          )}
          onClick={() => setPretty(!pretty)}
        >
          pretty print
        </code>
      </div>
      <code
        className={styles.code}
        style={{
          whiteSpace: pretty ? `pre` : `normal`,
        }}
      >
        {truncated && !showingAll
          ? `${text.slice(0, truncatedPos)}\n...`
          : text}
      </code>
      {showingAll ? (
        <code
          className={[styles.toggleShowAll, styles.active].join(` `)}
          onClick={() => {
            setShowingAll(!showingAll);
          }}
        >
          show less
        </code>
      ) : null}
    </div>
  );
}
