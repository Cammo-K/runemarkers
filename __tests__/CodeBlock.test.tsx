import '@testing-library/jest-dom';
import { act, fireEvent, render } from '@testing-library/react';
import CodeBlock from '@/components/atoms/CodeBlock';

const testText = `123456789`;

const testCodeBlock = (truncateLength: number | undefined) => (
  <CodeBlock truncateLength={truncateLength} text={testText} />
);

describe(`CodeBlock`, () => {
  it(`should render correctly`, () => {
    const { container } = render(testCodeBlock(5));
    expect(container).toMatchSnapshot();
  });

  it(`should not have the toggle if the text is short enough`, () => {
    const { container } = render(testCodeBlock(undefined));
    expect(container.querySelector(`#toggleShowAll`)).toBeFalsy();
  });

  it(`should show all on click`, () => {
    const { container } = render(testCodeBlock(5));
    const button = container.querySelector(`#toggleShowAll`);
    if (!button) {
      throw new Error(`No button found`);
    }
    act(() => {
      fireEvent.click(button);
    });
    expect(button.textContent).toBe(`Show Less`);
  });
});
