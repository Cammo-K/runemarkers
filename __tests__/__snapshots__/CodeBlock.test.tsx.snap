// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CodeBlock should render correctly 1`] = `
<div>
  <div
    class="container"
  >
    <h2
      class="header"
    >
      Tile Data:
    </h2>
    <pre
      style="display: block; overflow-x: auto; background: rgb(29, 31, 33); color: rgb(197, 200, 198); padding: 20px; border-radius: 10px; background-color: rgb(59, 59, 55); box-shadow: 0px 0px 10px rgba(0, 0, 0, 0.5); word-break: break-all;"
    >
      <code
        class="language-JSON"
        style="white-space: pre-wrap;"
      >
        <span>
          <span
            style="color: rgb(222, 147, 95);"
          >
            12345
          </span>
          <span>
            ...
          </span>
        </span>
      </code>
    </pre>
    <code
      id="toggleShowAll"
    >
      Show All (4 more characters, this may take a while to
        load)
    </code>
  </div>
</div>
`;
